cpp/daal/src/algorithms/dtrees/dtrees_train_data_helper.i

```diff
@@ -189,36 +204,109 @@ public:
     size_t size() const { return _aResponse.size(); }
     TResponse response(size_t i) const { return _aResponse[i].val; }
     const Response * responses() const { return _aResponse.get(); }
+
     bool reset(size_t n)
     {
         _aResponse.reset(n);
         return _aResponse.get() != nullptr;
     }
 
-    virtual bool init(const NumericTable * data, const NumericTable * resp, const IndexType * aSample)
+    bool resetWeight(size_t n)
     {
-        super::init(data, resp);
+        _aWeight.reset(n);
+        return _aWeight.get() != nullptr;
+    }
+
+    virtual bool init(const NumericTable * data, const NumericTable * resp, const IndexType * aSample, const NumericTable * weight = nullptr)
+    {
+        super::init(data, resp, weight);
         DAAL_ASSERT(_aResponse.size());
-        if (aSample)
+        if (_aWeight.get()) // use weight
         {
-            const IndexType firstRow = aSample[0];
-            const IndexType lastRow  = aSample[_aResponse.size() - 1];
-            ReadRows<algorithmFPType, cpu> bd(const_cast<NumericTable *>(resp), firstRow, lastRow - firstRow + 1);
-            const auto pbd = bd.get();
-            for (size_t i = 0; i < _aResponse.size(); ++i)
+            DAAL_ASSERT(_aWeight.size());
+            DAAL_ASSERT(_aWeight.size() == _aResponse.size());
+            if (aSample)
             {
-                _aResponse[i].idx = aSample[i];
-                _aResponse[i].val = TResponse(pbd[aSample[i] - firstRow]);
+                const IndexType firstRow = aSample[0];
+                const IndexType lastRow  = aSample[_aResponse.size() - 1];
+                ReadRows<algorithmFPType, cpu> bd(const_cast<NumericTable *>(resp), firstRow, lastRow - firstRow + 1);
+                const auto pbd = bd.get();
+                if (weight)
+                {
+                    ReadRows<algorithmFPType, cpu> bdw(const_cast<NumericTable *>(weight), firstRow, lastRow - firstRow + 1);
+                    const auto pbdw = bdw.get();
+                    for (size_t i = 0; i < _aResponse.size(); ++i)
+                    {
+                        _aResponse[i].idx = aSample[i];
+                        _aResponse[i].val = TResponse(pbd[aSample[i] - firstRow]);
+                        _aWeight[i].idx   = aSample[i];
+                        _aWeight[i].val =
+                            (isPositive<algorithmFPType, cpu>(TResponse(pbdw[aSample[i] - firstRow])) ? TResponse(pbdw[aSample[i] - firstRow]) :
```

Comment:
Do you follow SDL in the conversion into TResponse?

Reply:
It was fixed